src/Interpreters/Set.cpp

```diff
@@ -331,7 +348,16 @@ ColumnPtr Set::execute(const ColumnsWithTypeAndName & columns, bool negative) co
             result = castColumnAccurate(column_to_cast, data_types[i], cast_cache.get());
         }
 
-        materialized_columns.emplace_back() = result;
+        ColumnPtr col_to_emplace; /// If we cast DateTime64 column to other type, we lose its precision. if we have this case, we should not let this cast happen
+        if (isDateTime64(column_before_cast.column->getDataType()))
+            col_to_emplace = returnFilteredColumn(column_before_cast.column, res->getPtr());
```

Comment:
As far as I understand, it is always filled with zeros. Maybe we need result here?